File: extension/content/click_guard.js
// The submit denylist, mirrored from engine/autofill/click_guard.py.
// tests/test_extension_assets.py asserts DENY_TERMS here is term-for-term
// identical to DENY_TERMS there, so the two can never drift.
//
// The companion may click a field's own widget to set a value, but NEVER a
// control that submits/applies/advances/logs in. Every click in filler.js
// goes through jeClickGuard.isDenylisted first.
//
// Scope: judge the clicked element's OWN text/type/role + its DESCENDANTS,
// never its ancestors (so an option inside a form with a Submit is allowed,
// but a real submit button — or a wrapper containing one — is refused).
//
// Classic script: exposes window.jeClickGuard.
"use strict";

window.jeClickGuard = (function () {
  // Keep this list identical to click_guard.py DENY_TERMS (order-free).
  const DENY_TERMS = [
    "submit", "apply", "next", "continue", "save", "finish",
    "review and submit", "log in", "login", "sign in", "sign up",
    "register", "create account", "pay", "checkout", "proceed",
  ];

  function normalize(text) {
    return (text || "").trim().toLowerCase().replace(/\s+/g, " ");
  }

  // 019: the FINAL-class layer, mirrored from click_guard.py FINAL_TERMS
  // (parity-tested). A wizard's Continue is clickable under constitution
  // v1.2.0; the Submit at the end of it never is.
  const FINAL_TERMS = [
    "submit application", "submit my application", "review and submit",
    "submit", "create account", "create an account", "register",
    "sign up", "signup", "pay", "checkout", "place order",
    "confirm and submit",
  ];

  function isProgressionSafe(name) {
    const norm = normalize(name);
    if (!norm) { return false; }  // unnamed is not identifiably safe
    for (const term of FINAL_TERMS) {
      const re = new RegExp("(?<![a-z])"
        + term.replace(/[.*+?^${}()|[\]\\]/g, "\\$&") + "(?![a-z])");
      if (re.test(norm)) { return false; }
    }
    return true;
  }

  function isDenylistedSignal(text, type, role) {
    if ((type || "").trim().toLowerCase() === "submit") { return true; }
    const norm = normalize(text);
    if (!norm) { return false; }
    for (const term of DENY_TERMS) {
      const re = new RegExp("(?<![a-z])" + term.replace(/[.*+?^${}()|[\]\\]/g, "\\$&") + "(?![a-z])");
      if (re.test(norm)) { return true; }
    }
    return false;
  }

  // 019 (T036, FR-012): may the filler OPERATE this control to set a value?
  // Mirrors engine/autofill/click_guard.is_widget_operable. Text is judged
  // on the element's own accessible name — folding descendant text (the 011
  // rule, right for buttons) refused an ordinary dropdown whose surrounding
  // card contained the word "Next" or "Save". Descendant TYPES are still
  // folded: a wrapper containing a real submit control would submit the
  // form, and that is the danger the fold existed for.
  function ownName(el) {
    const aria = el.getAttribute && el.getAttribute("aria-label");
    if (aria) { return aria; }
    const labelled = el.getAttribute && el.getAttribute("aria-labelledby");
    if (labelled && window.jeScanner) {
      // reuse the scanner's resolver so both agree on what the name IS
      const root = el.getRootNode ? el.getRootNode() : document;
      const parts = labelled.split(/\s+/).map(function (id) {
        const node = (root.getElementById && root.getElementById(id))
          || document.getElementById(id);
        return node ? (node.innerText || node.textContent || "").trim() : "";
      }).filter(Boolean);
      if (parts.length) { return parts.join(" "); }
    }
    if (el.value) { return String(el.value); }
    // Own text only: direct text nodes, not descendants' text.
    let own = "";
    Array.prototype.forEach.call(el.childNodes || [], function (node) {
      if (node.nodeType === 3) { own += node.nodeValue; }
    });
    own = own.trim();
    if (own) { return own; }
    // A leaf element (an option row) has no element children — its
    // textContent IS its own name.
    return el.children && el.children.length === 0
      ? (el.textContent || "").trim() : "";
  }

  function isWidgetOperable(el) {
    if (!el) { return false; }
    const ownType = (el.getAttribute && el.getAttribute("type")) || el.type || "";
    if ((ownType || "").trim().toLowerCase() === "submit") { return false; }
    if (el.querySelector &&
        el.querySelector('button[type=submit], input[type=submit], [type=submit]')) {
      return false;
    }
    const role = ((el.getAttribute && el.getAttribute("role")) || "")
      .trim().toLowerCase();
    // An element the page declares an OPTION is a value, not a control:
    // "Next year" in a date menu is an answer. A submit dressed as an option
    // was already refused above, by type.
    if (role === "option") { return true; }
    return !isDenylistedSignal(ownName(el), ownType, role);
  }

  // Public: judge a DOM element (self + descendants).
  function isDenylisted(el) {
    if (!el) { return false; }
    const ownType = (el.getAttribute && el.getAttribute("type")) || el.type || "";
    const ownRole = (el.getAttribute && el.getAttribute("role")) || "";
    // own text + descendant text (textContent already includes descendants);
    // a descendant <button type=submit> makes the folded type "submit".
    let foldedType = ownType;
    if (el.querySelector) {
      const sub = el.querySelector('button[type=submit], input[type=submit], [type=submit]');
      if (sub) { foldedType = "submit"; }
    }
    const text = (el.textContent || el.value || el.getAttribute && el.getAttribute("aria-label") || "");
    return isDenylistedSignal(text, foldedType, ownRole);
  }

  return { DENY_TERMS, FINAL_TERMS, isDenylisted, isDenylistedSignal,
           isWidgetOperable, isProgressionSafe };
})();

File: engine/autofill/click_guard.py
"""The submit denylist (feature 011) — the single guarantee that the fill
engine, now allowed to click a field's own widget to set a value, NEVER
clicks a control that submits, applies, advances, saves, finishes, logs in,
registers, creates an account, or pays.

This module is the source of truth. `extension/content/click_guard.js`
mirrors `DENY_TERMS` term-for-term (a parity test in
tests/test_extension_assets.py fails if they drift). Both the Playwright
watcher and the browser companion consult the SAME term set at the moment of
clicking.

Scope (clarify Q1): the verdict is computed from the clicked element's OWN
text/type/role plus the text/type/role of elements it CONTAINS — never its
ancestors. So an option inside a form that also holds a Submit button is
allowed (the Submit is an ancestor), while a real submit button, or a
wrapper that contains one, is refused.
"""
from __future__ import annotations

import re

# Canonical submit-class terms (normalized, lowercase). Mirrored in JS.
# Order-independent; each is matched as a whole-phrase substring of the
# normalized element text.
DENY_TERMS: tuple[str, ...] = (
    "submit",
    "apply",
    "next",
    "continue",
    "save",
    "finish",
    "review and submit",
    "log in",
    "login",
    "sign in",
    "sign up",
    "register",
    "create account",
    "pay",
    "checkout",
    "proceed",
)


# 019 (T062, constitution v1.2.0): the FINAL-class layer. DENY_TERMS keeps
# protecting the fill path unchanged — it still refuses "next"/"continue"
# there, because a filler has no business advancing anything. Progression
# clicks are a separate, narrower permission, and these are the terms that
# permission never extends to. A wizard's Continue is now clickable; the
# Submit at the end of it is not, and never will be.
FINAL_TERMS: tuple[str, ...] = (
    "submit application",
    "submit my application",
    "review and submit",
    "submit",
    "create account",
    "create an account",
    "register",
    "sign up",
    "signup",
    "pay",
    "checkout",
    "place order",
    "confirm and submit",
)


def is_progression_safe(name: str = "") -> bool:
    """True when a control may be clicked to move the application FORWARD.

    False for anything that would submit it, create an account, or pay —
    the three things the human always does themselves (constitution v1.2.0).
    """
    norm = _normalize(name)
    if not norm:
        # An unnamed control is not identifiable as safe; refuse it.
        return False
    for term in FINAL_TERMS:
        if re.search(rf"(?<![a-z]){re.escape(term)}(?![a-z])", norm):
            return False
    return True


def _normalize(text: str) -> str:
    # lowercase, collapse whitespace, strip non-alphanumeric edges so
    # "Continue »" / "  submit  " normalize cleanly
    return re.sub(r"\s+", " ", (text or "").strip().lower())


def is_denylisted(text: str = "", type: str = "", role: str = "") -> bool:
    """True when this element must never be clicked. `text` is the element's
    (self+descendant) visible text; `type` its input/button type; `role` its
    ARIA role."""
    if (type or "").strip().lower() == "submit":
        return True
    norm = _normalize(text)
    if not norm:
        return False
    for term in DENY_TERMS:
        # whole-word/phrase match: the term bounded by non-letters (so
        # "apply" doesn't fire on "applied experience" mid-word, but does on
        # "apply now")
        if re.search(rf"(?<![a-z]){re.escape(term)}(?![a-z])", norm):
            return True
    return False


def is_widget_operable(own_name: str = "", descendant_text: str = "",
                       type_: str = "", role: str = "",
                       descendant_types: list[str] | None = None) -> bool:
    """019 (T036, FR-012): may the filler operate this control to SET A
    VALUE?

    The 011 rule folds descendant text into the verdict, which is right for
    judging a button but wrong for judging a widget: a react-select wrapper
    whose card happens to contain the words "Next" or "Save the date" was
    refused, so an ordinary dropdown became needs_manual. Text is therefore
    judged on the element's OWN accessible name.

    Descendant TYPES are still folded — a wrapper containing a real
    `type=submit` control would submit the form, and that danger is
    unchanged. `is_denylisted` itself is untouched: the fill path's
    button-judging behavior is exactly what it was.
    """
    types = [type_] + list(descendant_types or [])
    if any((t or "").strip().lower() == "submit" for t in types):
        return False
    # An element the page itself declares as an OPTION is a value, not a
    # control: "Next year" in a date menu, "Continue Education" in a
    # dropdown. Its text can therefore never make it submit-class — only a
    # submit type can, and that was refused above. (This is why
    # tests/fixtures/ats_pages/submit_styled_as_option.html exists: a submit
    # button dressed as an option is still caught, by type.)
    if (role or "").strip().lower() == "option":
        return True
    return not is_denylisted(own_name, type_, role)


def combined_signal(own_text: str = "", own_type: str = "", own_role: str = "",
                    descendant_texts: list[str] | None = None,
                    descendant_types: list[str] | None = None,
                    descendant_roles: list[str] | None = None) -> dict:
    """Fold the clicked element's own signal with its DESCENDANTS' (never
    ancestors) into a single {text, type, role} for is_denylisted. If any
    descendant is a submit-type control, the folded type is 'submit'."""
    texts = [own_text] + list(descendant_texts or [])
    types = [own_type] + list(descendant_types or [])
    folded_type = "submit" if any(
        (t or "").strip().lower() == "submit" for t in types
    ) else (own_type or "")
    return {
        "text": " ".join(t for t in texts if t),
        "type": folded_type,
        "role": own_role or "",
    }
